zproject/backends.py

```diff
@@ -208,3 +216,51 @@ class DevAuthBackend(ZulipAuthMixin):
 
     def authenticate(self, username, return_data=None):
         return common_get_active_user_by_email(username, return_data=return_data)
+
+class GitHubBackend(ZulipAuthMixin, GithubOAuth2):
+    def authenticate(self, *args, **kwargs):
+        try:
+            email_address = kwargs['response']['email']
+            return_data = kwargs['return_data']
+        except KeyError:
+            return None
+
+        try:
+            user_profile = get_user_profile_by_email(email_address)
+        except UserProfile.DoesNotExist:
+            return_data["valid_attestation"] = True
+            return None
+
+        if not user_profile.is_active:
+            return_data["inactive_user"] = True
+            return None
+
+        if user_profile.realm.deactivated:
+            return_data["inactive_realm"] = True
+            return None
+
+        return user_profile
+
+    def do_auth(self, *args, **kwargs):
+        # This function needs to be imported from here due to the cyclic
+        # dependency.
+        from zerver.views import login_or_register_remote_user
+
+        kwargs['return_data'] = {}
+
+        user_profile = super(GitHubBackend, self).do_auth(*args, **kwargs)
+
+        return_data = kwargs['return_data']
+        inactive_user = return_data.get('inactive_user')
+        inactive_realm = return_data.get('inactive_realm')
+
+        if inactive_user or inactive_realm:
+            return None
+
+        request = self.strategy.request
+        details = kwargs['response']
+        email_address = details.get('email')
+        full_name = details.get('name')
+
+        return login_or_register_remote_user(request, email_address,
+                                             user_profile, full_name)
```

Comment:
This looks reasonable.  

As a potential improvement, do you think there is a way we can refactor this so that it's easy to add another Social auth backend with just a few lines of code?  I'm imagining something like a `class SocialAuthMixin` and then being able to define `GitHubBackend` in terms of that.  Or if that doesn't work cleanly, we could at least move most of this code into helper functions so that GitHubBackend is <10 lines of code calling out to those.